makehelp.sh

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env bash
+# vim: ft=sh:sts=4:sw=4:expandtab
+
+AUTHOR="Ryan Volpe"
+VERSION="0.2"
+LICENSE="Apache Public License, v2.0"
+
+set -euo pipefail
+IFS=$'\n\t'
+
+STATIC_MODE=1
+MAKEFILE=
+
+while (( "$#" )); do
+    arg="${1:-}"
+    shift
+    case "$arg" in
+        --static)
+            STATIC_MODE=0
+            ;;
+
+        --help)
+            echo "Usage: $0 [OPTIONS] <Makefile>"
+            echo
+            echo -e "  Parses a \033[4mMakefile\033[0m for help documentation."
+            echo
+            echo "  This help documentation can be appended as a new make target via"
+            echo "     \`./makehelp.sh Makefile --static >> Makefile'."
+            echo "  The resulting Makefile will include help documentation without"
+            echo -e "  requiring distribution of \033[1mmakehelp\033[0m itself."
+            echo
+            echo "Options:"
+            echo "  --help   Show this message and exit."
+            echo "  --static Generate a new make target with a static help string."
+            exit 0
+            ;;
+
+        *)
+            MAKEFILE="$arg"
+            ;;
+    esac
+done
+
+if [ -z "$MAKEFILE" ]; then
+    echo "Error: a Makefile is required (see --help for details)" >&2
+    echo "Usage: $0 [OPTIONS] <Makefile>" >&2
+    exit 2
+fi
+
+if (( "$STATIC_MODE" == 0 )); then
+    echo ""
+    echo ".PHONY: makehelp"
+    echo "#: Display this help message and exit."
+    echo "makehelp:"
+    echo $'\t'"# Generated by makehelp.sh version $VERSION"
+    export _INJECT_MAKEHELP=0
+    output="$($0 $MAKEFILE)"
+    while read -r line; do
+        echo $'\t'"@echo '$line'"
+    done <<< "$output"
+    exit 0
+fi
+
+parse_makefile ()
+{
+    in_doc=1
+    decl=$'^([a-zA-Z0-9][^: ]*):'
+    declare -a buf
+    declare -a targets
+    src="$(grep -v -E $'^\t' "$1")"
+    for line in $src; do
+        stripped="${line##\#:}"
+        if (( ${#stripped} != ${#line} )); then
+            if (( $in_doc == 1 )); then
+                unset buf
+                declare -a buf
+                in_doc=0
+            fi
+            buf=( "${buf[@]}" "${stripped#"${stripped%%[![:space:]]*}"}" )
+        elif [[ "$line" =~ $decl ]]; then
+            if (( $in_doc == 0 )); then
+                printf "${BASH_REMATCH[1]}"
+                printf $'\1'
+                echo "${buf[@]:0}"
+                in_doc=1
+            fi
+            unset buf
+            declare -a buf
+        else
+            if (( $in_doc == 0 )); then
+                printf $'\1'
+                echo "${buf[@]:0}"
+                in_doc=1
+            fi
+            unset buf
+            declare -a buf
+        fi
+    done
+    if (( "${_INJECT_MAKEHELP:-}" == 0 )); then
```

Comment:
Currently this will inject a **makehelp** target into the output whether it's needed or not. While it would be odd for a Makefile to have a documented **makehelp** target and also run ``--static``, it's not unreasonable.

I don't have any better ideas at the moment, so will probably be opening a bug off of this.